plugins/oidc/access.go

```diff
@@ -10,118 +10,207 @@ import (
+var errDiscoveryUnmarshalFailure = errors.New("unable to unmarshal openid connect configuration")
+
+// discoverEndpoints retrieves the userinfo and jwks endpoints from the given
+// discovery endpoint.
+// The function sends a GET request to the discovery endpoint, which should
+// respond with the openid connect configuration.
+// This configuration is a JSON object that the function decodes into a map,
+// extracting the userinfo_endpoint and jwks_uri fields as the userinfo and jwks
+// endpoints respectively.
+// In case of error, it will also join the error with suitable contextual error
+// message and return.
+// If the discovery endpoint is empty or leads to a failed request, the function
+// returns an error.
+// If the retrieval or decoding of the openid connect configuration fails, the
+// function returns an error.
+//
+// Parameters:
+//
+//	discoveryEndpoint (string): address of the discovery endpoint.
+//
+// Returns:
+//
+//	userinfoEndpoint (string): the discovered userinfo endpoint.
+//	jwksEndpoint (string): the discovered jwks endpoint.
+//	err (error): the error, if any occurred during the process.
+func discoverEndpoints(discoveryEndpoint string) (userinfoEndpoint string, jwksEndpoint string, err error) {
+	// cleanup the discovery endpoint address and remove possible whitespaces
+	discoveryEndpoint = strings.TrimSpace(discoveryEndpoint)
+	if discoveryEndpoint == "" {
+		return "", "", errEmptyEndpoint
+	}
+
+	// now request the openid connect configuration endpoint
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
+}
+
+// extractBearerToken is a function that takes an authorization header value as
+// input and extracts the bearer token from it.
+// It uses a regular expression to match the header value against a predefined
+// pattern.
+// If a match is found and there are two capture groups in the match result, the
+// function returns the second capture group, which represents the bearer token.
+// If no match is found or there are not exactly two capture groups, an empty
+// string is returned.
+//
+// Example usage:
+//
+//	token := extractBearerToken("Bearer abcxyz123")
+//	// token = "abcxyz123"
+//
+// Parameters:
+//   - authorizationHeaderValue: The value of the authorization header.
+//
+// Returns:
+//
+//	The extracted bearer token, or an empty string if no valid bearer token is found.
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
```

Comment:
Why did you move the regex compile step into the function, will that be rerun every time?

Reply:
The compile step has been moved down here because sometimes the regex compilation failed. However, the regex compile step can be made to be only executed once, using the following code:
```go
// at the top of the file
const bearerTokenExtractionRegExString = `^(\w+) (\S+)$`
var bearerTokenExtractionRegEx *regexp.Regexp

func extractBearerToken(authorizationHeaderValue string) string {
    if bearerTokenExtractionRegEx == nil {
      bearerTokenExtractionRegEx = regexp.MustCompile(bearerTokenExtractionRegExString)
    }
    // rest of the function...
}
```
This way the regex will only compiled once, but only if the function is called

Reply:
This is a good alternative. But why would a hard-coded regex fail to compile sometimes?

Reply:
Local testing showed that not the regex failed to compile, but trying to use the regex resulted in a panic due to a nil pointer dereference. This happened if there was a request in the backlog of the gateway and the plugin is not started fully which happens if the configuration of the plugin has been changed. This is an edge case and was only reproduced by loading the gateway with about 25000 requests in its backlog and changing the gateway every 2.5 seconds.

Reply:
Huh, then we could just move the regex compile up again or not?

Reply:
Furthermore, the regex is now compiled only once and not at every access which reduces the time for handling the request